From 871d7874f9a7d37820d7f33e7a8f861354f22856 Mon Sep 17 00:00:00 2001
From: Liu Bolun <bolunx.liu@intel.com>
Date: Tue, 11 Sep 2012 11:29:48 +0800
Subject: [PORT FROM MAIN] Adding free heap when start encode fail.

BZ: 56455

we should hanlde error occur in initialization that calling
MediaSource::stop()when mState equals to LOADED_TO_IDLE,
it means OMXComponet is initializing only in initialization stage.
In addition, we not only stop mediasource, but also free buffer for
OMXCodec ports avoid memory leak.Modifying code style.

Change-Id: Ia7869bfd8f01da82d07502fab7956152350fd650
Orig-Change-Id: Ieb66b69d80f28a218d1cfabba9b9dac20fd35735
Signed-off-by: Liu Bolun <bolunx.liu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 media/libstagefright/OMXCodec.cpp |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 9da2042..18d575c 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -2572,6 +2572,16 @@ void OMXCodec::onEvent(OMX_EVENTTYPE event, OMX_U32 data1, OMX_U32 data2) {
         {
             CODEC_LOGE("ERROR(0x%08lx, %ld)", data1, data2);
 
+            // we should hanlde error occur in initialization that calling MediaSource::stop()
+            // when mState equals to LOADED_TO_IDLE, it means OMXComponet is initializing
+            // only in initialization stage.
+            if(mState == LOADED_TO_IDLE)
+            {
+               freeBuffersOnPort(kPortIndexInput);
+               freeBuffersOnPort(kPortIndexOutput);
+               mSource->stop();
+            }
+
             setState(ERROR);
             break;
         }
-- 
1.7.4.1

